File: Medium/199. Binary Tree Right Side View/rightSideView.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/*
    DFS解
    利用preorder的相反來訪問節點
    preorder的訪問順序為: 中左右
    我們可以變成中右左來確保先找rightmost的node，沒有的話在看左邊，都沒有的話代表這一層的右子樹是空的會去找左子樹
*/
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;
        dfs(root, 0, ans);
        return ans;
    }
private:
    void dfs(TreeNode* root, int d, vector<int>& ans){
        if(!root){
            return;
        }
        if(ans.size() <= d){
            ans.push_back(root->val);
        }
        dfs(root->right, d+1, ans);
        dfs(root->left, d+1, ans);
    }
};

File: Medium/199. Binary Tree Right Side View/rightSideView.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        doDFS(root, 0, ans);
        return ans;
    }
    private void doDFS(TreeNode root, int d, List<Integer> ans) {
        if(root == null){
            return;
        }
        if(ans.size() <= d){
            ans.add(root.val);
        }
        doDFS(root.right, d+1, ans);
        doDFS(root.left, d+1, ans);
    }
}